### PR TITLE
Feature/add log grouping to UI

### PR DESCRIPTION
**Pull request by jscheffl**
This PR is another "eye candy" for upcoming 2.9.0 release - and I hope I pass the review and merge if before release cut-off.

Logging is usually a mass of text. Tasks/Operators generate a lot of text and it is hard to read and understand what is going on. Expertise need to be built-up. In our departments users are mostly over-whelmed with the logs in Airflow.

A bit of filtering had been introduced but what I personally always was missing was a kind of log grouping which is existing in [Github](https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#grouping-log-lines) or also [Azure DevOps Pipelines](https://learn.microsoft.com/en-us/azure/devops/pipelines/scripts/logging-commands?view=azure-devops&tabs=powershell#formatting-commands).

This PR adds Log Grouping also to Airflow with the following effects:
- Log grouping can be made with the same keywords like in Github or AzureDevOps - means that if users have scripts from these systems, they could be used in Airflow as well. Start a group with `::group::Some Headline` and end it with `::endgroup::`
- Task instance runs automatically "fold" log details of task execution prior and post `execute()` method, making logs clean per default (showing only real business logic)
- If a user un-folds a section, un-folding state is remembered in Grid view
- Grouping is also working in old/legacy log view as well as Grid if the log is updated (auto-tail when task is running)

Example, how do logs look like for the `example_params_trigger_ui` DAG:
image

If you click on the blue colored log group header, the section un-folds:
image

One example added to `example_python_operator` DAG, prints are grouped giving the following results in the task logs:
image

FYI @clellmann @wolfdn @AutomationDev85

### CODE REVIEW

**Comment by potiuk** [comment]
I ❤️ it. Been using it in GitHub for ages.

Folding before/after execute automatically is cool.

**Comment by potiuk** [comment]
Just a warning - there **might** be some non-trivial performance issue created by it. But I think it's worth risking it.

**Comment by jscheffl** [comment]
> Just a warning - there **might** be some non-trivial performance issue created by it. But I think it's worth risking it.

Yeah, mainly depending on your browser/CPU speed and log volume. Anyway will only hit of the regex can make a match in log lines, but will need to crawl over N log lines. Will do a short test. Today log is stopping to be processed at 1M lines, asking for download.

**Comment by jscheffl** [comment]
> Just a warning - there **might** be some non-trivial performance issue created by it. But I think it's worth risking it.

Okay, did a test with a DAG and just created ~40.000 and ~400.000 log lines:
- 40k lines: loading time 1-2s in grid view
- 400k lines: loading time of 12-15s, then scrolling and viewing/folding/unfolding was 100% smooth
  - View of same logs in "Legacy" non Grid foll-screen log view took only 6s!

Test on Core i7-1280P / HP Laptop / Firefox (other browsers might be even faster)
Test on Airflow running via breeze, NOT dev mode (optimized JS compiled!)
Note that production of 400k print lines of LOG in Python 3.11 took 55 seconds to execute - viewing logs in browser after 15s was quite fast for this :-D - Log file was in total 32MB for 400.000 lines

**Review by hussein-awala — COMMENTED** [review]
Nice one, it looks good!

I agree with Jarek on the performance issue, but IMHO we can merge it into beta and then decide if we need to add configuration to enable the feature and disable it by default.

File: airflow/www/static/js/dag/details/taskInstance/Logs/utils.ts [hunk 1]
```diff
@@ -62,6 +63,16 @@ export const parseLogs = (
   const fileSources: Set<string> = new Set();
   const ansiUp = new AnsiUp();
 
+  const urlRegex = /((https?:\/\/|http:\/\/)[^\s]+)/g;
+  // Detect log groups which can be collapsed
+  // Either in Github like format '::group::<group name>' to '::endgroup::'
+  // see https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#grouping-log-lines
+  // Or in ADO pipeline like format '##[group]<group name>' to '##[endgroup]'
+  // see https://learn.microsoft.com/en-us/azure/devops/pipelines/scripts/logging-commands?view=azure-devops&tabs=powershell#formatting-commands
+  const logGroupStart = / INFO - (::|##\[])group(::|\])([^\n])*/g;
```
Comment (bbovenzi) [inline airflow/www/static/js/dag/details/taskInstance/Logs/utils.ts:72]: Log groups are only supported for INFO logs?

File: airflow/www/static/js/dag/details/taskInstance/Logs/utils.ts [hunk 2]
```diff
@@ -103,10 +114,27 @@ export const parseLogs = (
...
+          const gName = textLine.substring(17);
+          const gId = gName.replace(/\W+/g, "_").toLowerCase();
+          if (unfoldedLogGroups.indexOf(gId) === -1) {
+            const unfold = `<span id="${gId}_unfold" style="${logGroupStyle}"> &#11208; ${gName}</span>`;
+            const fold = `<span style="display:none;"><span id="${gId}_fold" style="${logGroupStyle}"> &#11206; ${gName}</span>`;
+            return unfold + fold;
+          }
+          const unfold = `<span id="${gId}_unfold" style="display:none;${logGroupStyle}"> &#11208; ${gName}</span>`;
+          const fold = `<span><span id="${gId}_fold" style="${logGroupStyle}"> &#11206; ${gName}</span>`;
+          return unfold + fold;
```
Comment (bbovenzi) [inline airflow/www/static/js/dag/details/taskInstance/Logs/utils.ts:133]: There is a lot of repeated code here. We're only changing `display:none;` So let's just have something like:

```
const isFolded = unfoldedLogGroups.indexOf(gId) === -1;
const unfoldStyle = isFolded ? '' : 'display:none;'
const foldStyle = isFolded ? 'display:none;' : '';
```

File: airflow/www/static/js/dag/details/taskInstance/Logs/utils.ts [hunk 3]
```diff
@@ -62,6 +63,16 @@ export const parseLogs = (
...
 
+  const urlRegex = /((https?:\/\/|http:\/\/)[^\s]+)/g;
+  // Detect log groups which can be collapsed
+  // Either in Github like format '::group::<group name>' to '::endgroup::'
+  // see https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#grouping-log-lines
+  // Or in ADO pipeline like format '##[group]<group name>' to '##[endgroup]'
+  // see https://learn.microsoft.com/en-us/azure/devops/pipelines/scripts/logging-commands?view=azure-devops&tabs=powershell#formatting-commands
+  const logGroupStart = / INFO - (::|##\[])group(::|\])([^\n])*/g;
+  const logGroupEnd = / INFO - (::|##\[])endgroup(::|\])/g;
+  const logGroupStyle = "color:#0060df;cursor:pointer;font-weight: bold;";
```
Comment (bbovenzi) [inline airflow/www/static/js/dag/details/taskInstance/Logs/utils.ts]: Could you leave a comment on what color this hex is?

File: airflow/www/static/js/dag/details/taskInstance/Logs/LogBlock.tsx [hunk 4]
```diff
@@ -59,10 +67,43 @@ const LogBlock = ({ parsedLogs, wrap, tryNumber }: Props) => {
     }
   };
 
+  const onClick = (e: React.MouseEvent<HTMLElement>) => {
+    const target = e.target as HTMLElement;
```
Comment (bbovenzi) [inline airflow/www/static/js/dag/details/taskInstance/Logs/LogBlock.tsx:71]: const onClick = ({ currentTarget }: React.MouseEvent<HTMLSpanElement>) => {
    const target = currentTarget;
    
We can get ride of a few `as` castings.

Comment (bbovenzi) [inline airflow/www/static/js/dag/details/taskInstance/Logs/LogBlock.tsx:71]:
```suggestion
  const onClick = ({ currentTarget }: React.MouseEvent<HTMLSpanElement>) => {
    const target = currentTarget;
```
We can get rid of some `as` castings

File: airflow/www/static/js/dag/details/taskInstance/Logs/LogBlock.tsx [hunk 5]
```diff
@@ -59,10 +67,43 @@ const LogBlock = ({ parsedLogs, wrap, tryNumber }: Props) => {
...
+      // remember the folding state if logs re-loaded
+      if (unfoldedGroups.indexOf(gId) >= 0) {
+        unfoldedGroups.splice(unfoldedGroups.indexOf(gId), 1);
+      }
+      setUnfoldedLogGroup(unfoldedGroups);
+      // now do the folding
+      if (target.parentNode) {
+        (target.parentNode as HTMLElement).style.display = "none";
+        if (target.parentNode.previousSibling) {
+          (target.parentNode.previousSibling as HTMLElement).style.display =
```
Comment (bbovenzi) [inline airflow/www/static/js/dag/details/taskInstance/Logs/LogBlock.tsx]:
```suggestion
      if (target.parentElement) {
        target.parentElement.style.display = "none";
        if (target.parentElement.previousSibling) {
          (target.parentElement.previousSibling as HTMLElement).style.display =
```

File: airflow/www/static/js/dag/details/taskInstance/Logs/LogBlock.tsx (hunk 4 above)
Comment (jscheffl) [reply airflow/www/static/js/dag/details/taskInstance/Logs/LogBlock.tsx:71]: I tried to apply this but it seems from typing perspective `currentTarget` is different than `target`. Here I need `target`to know which element was clicked in the log frame, not the log frame itself.

`target` is of type `EventTarget`
`currentTarget` is of type `EventTarget & HTMLElement`

I tried multiple options but with `target` I get errors in pre-commit like:
```
static/js/dag/details/taskInstance/Logs/LogBlock.tsx:71:18 - error TS2339: Property 'id' does not exist on type 'EventTarget'.

71     alert(target.id?.endsWith("_unfold"));
                    ~~

static/js/dag/details/taskInstance/Logs/LogBlock.tsx:72:16 - error TS2339: Property 'id' does not exist on type 'EventTarget'.

72     if (target.id?.endsWith("_unfold")) {
                  ~~

static/js/dag/details/taskInstance/Logs/LogBlock.tsx:73:26 - error TS2339: Property 'id' does not exist on type 'EventTarget'.

73       const gId = target.id.substring(0, target.id.length - 7);
                            ~~

static/js/dag/details/taskInstance/Logs/LogBlock.tsx:73:49 - error TS2339: Property 'id' does not exist on type 'EventTarget'.

73       const gId = target.id.substring(0, target.id.length - 7);
                                                   ~~

static/js/dag/details/taskInstance/Logs/LogBlock.tsx:78:14 - error TS2339: Property 'style' does not exist on type 'EventTarget'.

78       target.style.display = "none";
                ~~~~~

static/js/dag/details/taskInstance/Logs/LogBlock.tsx:79:18 - error TS2339: Property 'nextElementSibling' does not exist on type 'EventTarget'.

79       if (target.nextElementSibling) {
                    ~~~~~~~~~~~~~~~~~~

static/js/dag/details/taskInstance/Logs/LogBlock.tsx:80:17 - error TS2339: Property 'nextElementSibling' does not exist on type 'EventTarget'.

80         (target.nextElementSibling as HTMLElement).style.display = "inline";
                   ~~~~~~~~~~~~~~~~~~

static/js/dag/details/taskInstance/Logs/LogBlock.tsx:83:16 - error TS2339: Property 'id' does not exist on type 'EventTarget'.

83     if (target.id?.endsWith("_fold")) {
                  ~~

static/js/dag/details/taskInstance/Logs/LogBlock.tsx:84:26 - error TS2339: Property 'id' does not exist on type 'EventTarget'.

84       const gId = target.id.substring(0, target.id.length - 5);
                            ~~

static/js/dag/details/taskInstance/Logs/LogBlock.tsx:84:49 - error TS2339: Property 'id' does not exist on type 'EventTarget'.

84       const gId = target.id.substring(0, target.id.length - 5);
                                                   ~~

static/js/dag/details/taskInstance/Logs/LogBlock.tsx:91:18 - error TS2339: Property 'parentElement' does not exist on type 'EventTarget'.

91       if (target.parentElement) {
                    ~~~~~~~~~~~~~

static/js/dag/details/taskInstance/Logs/LogBlock.tsx:92:16 - error TS2339: Property 'parentElement' does not exist on type 'EventTarget'.

92         target.parentElement.style.display = "none";
                  ~~~~~~~~~~~~~

static/js/dag/details/taskInstance/Logs/LogBlock.tsx:93:20 - error TS2339: Property 'parentElement' does not exist on type 'EventTarget'.

93         if (target.parentElement.previousSibling) {
                      ~~~~~~~~~~~~~

static/js/dag/details/taskInstance/Logs/LogBlock.tsx:94:19 - error TS2339: Property 'parentElement' does not exist on type 'EventTarget'.

94           (target.parentElement.previousSibling as HTMLElement).style.display =
                     ~~~~~~~~~~~~~


Found 14 errors in the same file, starting at: static/js/dag/details/taskInstance/Logs/LogBlock.tsx:71
```

File: airflow/www/static/js/dag/details/taskInstance/Logs/utils.ts (hunk 3 above)
Comment (jscheffl) [reply airflow/www/static/js/dag/details/taskInstance/Logs/utils.ts]: Actually I did not want to hard-code a color here - I just selected the same color like we have with other areas of "blue". Added a comment - if there is a proper/better was to use CSS let me know. Could not find any reasonable way but would like to also offer making it better.

**Comment by jscheffl** [comment]
Did some further performance measurement on my Laptop. Interesting, if logs are folded actually rendering is even faster than if no grouping... maybe because most content is hidden and saves rendering time:

|  | main before log coloring PR #37985 | main today | this PR |
|--|-- | --|--|
| Log creation with DAG|-|-|43s|
| Firefox - Logs in Grid View | 20s |18.2s|15.9s| 
| Firefox - Logs in Legacy View | 36.3|37.1|14.8|
| Chromium - Logs in Grid View|21.3s|20.5s|14.8s|
| Chromium - Logs in Legacy View|24.3s|25.3s|15.2s|

All measured with my HP Laptop, Intel Core i7-1280P, Ubuntu 20.04. Measurement w/o proper tooling, just with a stop-watch :-D

FYI the 400k logs were generated with the following code in a PythonOperator:
```
        print("Now some log-mess...")
        for i in range(400):
            print(f"::group::Mess #{i}")
            for j in range(1000):
                print(f"messi line {i} - {j}")
            print("::endgroup::")
        print("DONE with log-mess!")
```

**Comment by potiuk** [comment]
> Did some further performance measurement on my Laptop. Interesting, if logs are folded actually rendering is even faster than if no grouping... maybe because most content is hidden and saves rendering time:

😱  -> yeah. that's likely not a very "realistic" example, but good enough. From my experience in GitHub Actions, the biggest issue is scrolling through HUGHE folded group - but I think a lot of it is becuase of the relatively slow backend handling a LOT LOT LOT of clients vs relatively fast front-end. Airlfow has far less of that problem - the webserver will be handling few (maybe few tens) of people maximum.

**Comment by jscheffl** [comment]
> > Did some further performance measurement on my Laptop. Interesting, if logs are folded actually rendering is even faster than if no grouping... maybe because most content is hidden and saves rendering time:
> 
> 😱 -> yeah. that's likely not a very "realistic" example, but good enough. From my experience in GitHub Actions, the biggest issue is scrolling through HUGHE folded group - but I think a lot of it is becuase of the relatively slow backend handling a LOT LOT LOT of clients vs relatively fast front-end. Airlfow has far less of that problem - the webserver will be handling few (maybe few tens) of people maximum.

Agree, not realistic but wanted to see the drop - and was surprised it is even faster in the case. I was fearing it is a lot slower but I have no fear anymore. ~😱~

File: airflow/www/static/js/dag/details/taskInstance/Logs/LogBlock.tsx (hunk 4 above)
Comment (bbovenzi) [reply airflow/www/static/js/dag/details/taskInstance/Logs/LogBlock.tsx:71]: Ok we can leave it as is then

File: airflow/www/static/js/dag/details/taskInstance/Logs/LogBlock.tsx [hunk 6]
```diff
@@ -59,10 +67,43 @@ const LogBlock = ({ parsedLogs, wrap, tryNumber }: Props) => {
     }
   };
 
+  const onClick = (e: React.MouseEvent<HTMLElement>) => {
+    const target = e.target as HTMLElement;
+    if (target.id?.endsWith("_unfold")) {
+      const gId = target.id.substring(0, target.id.length - 7);
```
Comment (bbovenzi) [inline airflow/www/static/js/dag/details/taskInstance/Logs/LogBlock.tsx]: I'm not the biggest fan of just hard-coding the number `7`. Perhaps we could make `const unfoldString = `_unfold` and we can use `unfoldString.length`. Same for fold

Comment (jscheffl) [reply airflow/www/static/js/dag/details/taskInstance/Logs/LogBlock.tsx]: Good point. Actually very good point. Adjusted in all places with constants.

**Review by bbovenzi — APPROVED** [review]
Great work!

**Comment by dstandish** [comment]
One issue with this is (IIUC) that it assumes that groups won't overlap.  But logs for a TI can come from multiple sources and this might not always be true.  First it was only worker.  Then we added triggerer.  Then we added scheduler (in exceptional circumstances).  Now there is interest in logging to TI logs any action specific to a TI. (e.g. scheduler state changes).  With AIP-44, there could be logs from webserver / internal API server.  To the extent that these groups can overlap in time, the folding will be misleading.

**Comment by jscheffl** [comment]
> One issue with this is (IIUC) that it assumes that groups won't overlap. But logs for a TI can come from multiple sources and this might not always be true. First it was only worker. Then we added triggerer. Then we added scheduler (in exceptional circumstances). Now there is interest in logging to TI logs any action specific to a TI. (e.g. scheduler state changes). With AIP-44, there could be logs from webserver / internal API server. To the extent that these groups can overlap in time, the folding will be misleading.

Yes, agree - partly. Groups can overlap. And if you have a large codebase you need to be careful. But you are not focred to use grouping in your custom code of tasks. I hope I took care in the case in pre/post execution that I catched all cases. If you thing explicitly I missed any let me know. Happy to add a patch/bugfix.

As of your concerns I just added an explicit check which I did not do before... what happens if you nest groups...?!?
Experiment with group nesting:
```
        print("before")
        print("::group::Outer Group")
        print("within 1")
        print("::group::Inner Group 1")
        print("within 2a - inside a second group")
        print("::endgroup::")
        print("within 2b")
        print("::group::Inner Group 2")
        print("within 2c - inside a second group")
        print("::endgroup::")
        print("within 3")
        print("::endgroup::")
        print("after")
```
Result: Seems to work. Nested groups allow nested folding:
image

If the numbers of group starts and ends do not match logs might look "weired". I'd assume if more groups are opened than ended still you can un-fold. Too many group ends will not be a problem. If group starts and ends are completely mixed the context might look strange but in general I'd assume all logs should be at least un-foldable by the user.

If you for example take this bad usage:
```
        print("before")
        print("::group::Group 1 forgot to end")
        print("within 1")
        print("::group::Group 2 forgot to end")
        print("within 2")
        print("::group::Group 3 forgot to end")
        print("within 3")
        print("after")
```
Result will render a folding group to the end of logs "eating up" other groups:
image

...but still it is possible to un-fold all:
image

**Pull request merged**